plugins/org.polarsys.capella.transition.system2subsystem/src/org/polarsys/capella/transition/system2subsystem/rules/fa/FunctionalChainInvolvementRule.java

```diff
@@ -406,23 +347,13 @@ protected void premicesRelated(EObject element_p, ArrayList<IPremise> needed_p)
     if (element_p instanceof FunctionalChainInvolvement) {
       FunctionalChainInvolvement src = (FunctionalChainInvolvement) element_p;
 
-      Collection<EObject> nexts = new ArrayList<EObject>();
-
-      while ((src.getNextFunctionalChainInvolvements() != null)
-          && (src.getNextFunctionalChainInvolvements().isEmpty() == false)) {
-        src = src.getNextFunctionalChainInvolvements().get(0);
-        nexts.add(src);
-
-      }
-      needed_p.addAll(createDefaultPrecedencePremices(nexts, "nextsInvolvements"));
-    }
-
-    if (element_p instanceof FunctionalChainReference) {
-      Collection<EObject> fcis = new ArrayList<EObject>();
-      FunctionalChain fc = ((FunctionalChainReference) element_p).getReferencedFunctionalChain();
-      fcis.addAll(fc.getOwnedFunctionalChainInvolvements());
+      FunctionalChainAttachmentHelper helper = FunctionalChainAttachmentHelper.getInstance(getCurrentContext());
+      needed_p.addAll(createDefaultPrecedencePremices((Collection) helper.getNextValid(src, getCurrentContext()),
+          "nextValidInvolvements"));
 
-      needed_p.addAll(createDefaultCriticalPremices(fcis, "referencedInvolvements"));
+      Set<FunctionalChainReference> refs = getPaths((FunctionalChainInvolvement) element_p).stream()
```

Comment:
but we rather need that all using references are transformed as we will add them in targetHierarchy/sourceHierarchy